hibernate-core/src/main/java/org/hibernate/event/spi/EventType.java

```diff
@@ -76,6 +78,41 @@
 	public static final EventType<PostCollectionRemoveEventListener> POST_COLLECTION_REMOVE = create( "post-collection-remove", PostCollectionRemoveEventListener.class );
 	public static final EventType<PostCollectionUpdateEventListener> POST_COLLECTION_UPDATE = create( "post-collection-update", PostCollectionUpdateEventListener.class );
 
+	/**
+	 * Add a new event type.
+	 *
+	 * @param name - name of the custom event
+	 * @param listenerClass - the base listener class or interface associated with the entity type
+	 * @param <T> - listenerClass
+	 * @return the custom {@link EventType}
+	 */
+	public synchronized static <T> EventType<T> addCustomEventType(String name, Class<T> listenerClass) {
```

Comment:
The `EVENT_TYPE_BY_NAME_MAP` is not safe for concurrent access; I see you added a `synchronized` on this method, which is correct as it performs write operations on the map.

But this implies that we'll need to have the same lock on reads as well; perhaps the map should be changed from `HashMap` to a `ConcurrentHashMap`?

Alternatively, I wonder if the map could be statically initialized by discovering a service. If you define a "CustomEventTypeContributor" interface, then Hibernate RX could implement it and you could load them when this class is initialized.

Reply:
I original made this method `synchronized` because I had added a `List` for resolving an `EventType` by `ordinal` for an earlier incarnation of the fix. This was necessary to make sure that both `EVENT_TYPE_BY_NAME_MAP` and the `List` were kept in sync. I ended up not needing the `List`, so I deleted it. 

I thought it was still a good idea to keep this method method `synchronized`. I didn't really see a reason for changing the `HashMap` to `ConcurrentHashMap`. It seemed very unlikely that an application would look up an `EventType` by name before creating the custom `EventType` via `EventType#addCustomEventType`. I suppose it is possible though.

I'll do as you suggest: 1) remove `synchronized`; 2) change `HashMap` to  `ConcurrentHashMap`.

Reply:
Regarding:

> Alternatively, I wonder if the map could be statically initialized by discovering a service. If you define a "CustomEventTypeContributor" interface, then Hibernate RX could implement it and you could load them when this class is initialized.

It could be done that way, but I'm not sure it's worthwhile. hibernate-rx already has an implementation of `Integrator` for registering listeners: [org.hibernate.rx.service.RxIntegrator](https://github.com/hibernate/hibernate-rx/blob/master/hibernate-rx-core/src/main/java/org/hibernate/rx/service/RxIntegrator.java). 

It would be simple for `Integrator#integrate` to add the custom `EventType` before adding its corresponding listeners.

I could go either way on this. 

@sebersole, do you have an opinion on this?

Reply:
> It seemed very unlikely that an application would look up an EventType by name before creating the custom EventType via EventType#addCustomEventType. I suppose it is possible though.

Keep in mind that the problem is not strictly about the order in which such operations might happen: the problem lies in the visibility rules. Even read operations need to ensure that any state change which has happened (at any point in the past) is actually "made visible" to them.

> I'll do as you suggest: 1) remove synchronized; 2) change HashMap to ConcurrentHashMap.

+1 to use a ConcurrentMap, but please keep the `synchronized` : like @NathanQingyangXu mentions, otherwise there's a possible race condition.

Thanks!

Reply:
@NathanQingyangXu, I believe I have dealt with your first point (`f (!map.containsKey(key)) {...}` in my most recent commit.

Reply:
> @sebersole, do you have an opinion on this?

@gbadner I prefer dedicated integration contracts (`TypeContributor`, e.g.) rather than generic ones (`Integrator`).  Here, I'd actually even consider defining a contract for listener registration as well rather than relying on the generic `Integrator`.  Something like:

```
interface EventSystemContributions {
    <T> void registerEventType(EventType<T> type);

    <T> void prependListener(EventType<T> type, T listener);
    <T> void appendListener(EventType<T> type, T listener);

    void registerDuplicationStrategy(DuplicationStrategy strategy);
    void registerDuplicationStrategy(EventType type, DuplicationStrategy strategy);
}

interface EventSystemContributor {
    void contribute(EventSystemContributions target);
}
```

Reply:
I mentioned to Gail that I started a PR of my own on top of hers to look at my suggestion.  I wonder what y'all think about some related clean-up...

The main problem I had is with keeping track of custom event types on EventType itself.  Ugly chicken/egg situation that requires way too much static state.  At a super high-level I'd suggest instead to keep `EventType` simple - just the class definition and the static standard type declarations.  In fact, in retrospect, knowing that customer event-types were in the future I would have split even those out into a different class `StandardEventTypes` e.g.  But specifically here, I want to get rid of anything related to custom types.

Any customization is logically part of the registry, not the EventType class.  As such, imo, it belongs as part of the registry.  

All together this has the benefit of removing the worries about static wrt entries "bleeding" across VMs.

Anyway, I'm going to take that approach in my PR.  The "downside" is that certain methods like `EventType#values` no longer has the same meaning.  Well, it does... it has the same meaning it had before any of these "custom type" changes - it would simply be the list of standard types.